keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots/README.md

```diff
@@ -0,0 +1,1246 @@
+	// If APIGroup is not specified, the specified Kind must be in the core API group.
+	// For any other third-party types, APIGroup is required.
+	// +optional
+	APIGroup *string
+	// Kind is the type of resource being referenced
+	Kind string
+	// Namespace is the namespace of resource being referenced
+	// Note that when a namespace is specified, a ReferenceGrant object is required in the referent namespace to allow that namespace's owner to accept the reference. See the ReferenceGrant documentation for details.
+	// +optional
+	Namespace string
+	// Name is the name of resource being referenced
+	Name string
+}
+```
+
+### Implementation
+
+#### CSI external provisioner
+
+- If CrossNamespaceSourceProvisioning feature gate isn't enabled: Do the existing provisioner behavior.
+- Otherwise: Datasource is handled in the below reconciler loop:
+  - If data source is specified through `DataSource` or `DataSourceRef`: Do the existing provisioner behavior for the data source.
+  - If data source is specified through `DataSourceRef2`:
+    - If namepace isn't specified: Provision a volume from the snapshot specified through DataSourceRef2`.
+    - Otherwise:
+      - If the access to the `VolumeSnapshot` referenced by the `PersistentVolumeClaim` is allowed by any `ReferenceGrant`s: Provision a volume from the snapshot specified through DataSourceRef2`.
+      - Otherwise: Do nothing (the `ReferenceGrant` check above will be repeated in the loop).
+
+#### API server
+
+To provide compatibility of `PersistentVolumeClaim` API with older versions, the API servers need to properly handle `DataSourceRef2`, `DataSource`, and `DataSourceRef`.
+There will be two approaches:
+
+1. Eventually allow only `DataSourceRef2` fields
+
+This is [a similar approach](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators#proposal) to `DataSourceRef` and `DataSource` compatibility handling.
+
+Deprecate the DataSourceRef field once DataSourceRef2 reaches GA, but continue to support it in a backwards-compatible way, never removing it.
+
+In particular:
+  1. If DataSourceRef and DataSourceRef2 are both unset; accept
+  2. If DataSourceRef is set AND DataSourceRef2 is not set; set DataSourceRef2 from DataSourceRef.
+  3. If DataSourceRef is set AND DataSourceRef2 is set the same (namespace is empty); accept
+  4. If DataSourceRef is set AND DataSourceRef2 is set differently; reject
+  5. If DataSourceRef is not set AND DataSourceRef2 is set (namespace is empty); set DataSourceRef from DataSourceRef2 and accept
+  6. If DataSourceRef is not set AND DataSourceRef2 is set (namespace is not empty); accept
+
+Note that we can't set DataSourceRef from DataSourceRef2 in case 6, because namespace field can't be set to DataSourceRef.
+Impact of not setting fields in this case, needs to be considered.
+
+2. Allow both `DataSourceRef` field and `DataSourceRef2` fields to coexist, but only allow specifying either of them or not specifying at all
```

Comment:
I think we should start with this approach initially. Once the ReferenceGrant model has been more proven then I think we can consider merging the two.